service/hproxy/hproxy.go

```diff
@@ -0,0 +1,960 @@
+//	//}
+//
+//	// log.Errorf("proxy error: %T", e)
+//	// log.Errorf("proxy error: %v", cs.Sdump(e))
+//
+//	// connection errors
+//	// XXX or is this all ohio?
+//	var netErr net.Error
+//	switch {
+//	case errors.As(e, &netErr) && netErr.Timeout():
+//		w.WriteHeader(http.StatusBadGateway)
+//	case errors.Is(e, net.ErrClosed):
+//		w.WriteHeader(http.StatusBadGateway)
+//	case errors.Is(e, net.ErrWriteToConnected):
+//		w.WriteHeader(http.StatusBadGateway)
+//	case errors.Is(e, syscall.ECONNREFUSED):
+//		w.WriteHeader(http.StatusBadGateway)
+//	default:
+//		panic(e)
+//	}
+//}
+
+func (s *Server) _clientRemove(remoteAddr string) {
+	if v, ok := s.clients[remoteAddr]; ok {
+		v.abort()
+		delete(s.clients, remoteAddr)
+	}
+}
+
+func (s *Server) clientRemove(remoteAddr string) {
+	s.mtx.Lock()
+	s._clientRemove(remoteAddr)
+	s.mtx.Unlock()
+}
+
+func (s *Server) nodeAdd(node string) error {
+	u, err := url.Parse(node)
+	if err != nil {
+		return fmt.Errorf("invalid url: %w", err)
+	}
+	switch u.Scheme {
+	case "http", "https":
+	default:
+		return fmt.Errorf("unsuported scheme: %v", u.Scheme)
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	// make sure it isn't a dupe
+	for k := range s.hvmHandlers {
+		if s.hvmHandlers[k].u.String() == u.String() {
+			if s.hvmHandlers[k].state == StateRemoved {
+				s._nodeNew(k) // add it back
+				return nil
+			}
+			return errors.New("duplicate")
+		}
+	}
+
+	hvmHandler := HVMHandler{
+		id:    len(s.hvmHandlers),
+		state: StateNew,
+		u:     u,
+		c: &http.Client{
+			Transport: &http.Transport{
+				DialContext:           dialer,
+				TLSHandshakeTimeout:   5 * time.Second,
+				ResponseHeaderTimeout: s.cfg.RequestTimeout,
+			},
+		},
+		// Ethereum only needs what is set. If we want to make this
+		// generic we may have to fart with this a bit.
+		rp: &httputil.ReverseProxy{
+			Rewrite: func(r *httputil.ProxyRequest) {
+				r.SetURL(u)
+				r.SetXForwarded()
+			},
+			Director: nil, // not needed
+			Transport: &http.Transport{
+				DialContext:           dialer,
+				TLSHandshakeTimeout:   5 * time.Second,
+				ResponseHeaderTimeout: s.cfg.RequestTimeout,
+				MaxIdleConnsPerHost:   1,
+				MaxConnsPerHost:       1,
+			},
+			FlushInterval:  0,
+			ErrorLog:       nil, // XXX wrap in loggo
+			BufferPool:     nil, // not useful for different sized calls
+			ModifyResponse: nil, // not needed
+			ErrorHandler:   nil, // s.handleProxyError, ModifyResponse only
+		},
+	}
+
+	// For now, everything is ethereum but we can use the poker function to
+	// handle different types health checks.
+	hvmHandler.poker = NewEthereumProxy(func(ctx context.Context) error {
+		resp, err := CallEthereum(ctx, hvmHandler.c, hvmHandler.u.String(),
+			"eth_blockNumber", nil)
+		if err != nil {
+			return err
+		}
+
+		j := make(map[string]any)
+		err = json.Unmarshal(resp, &j)
+		if err != nil {
+			return err
+		}
+
+		// Make sure we have "result"
+		if _, ok := j["result"]; !ok {
```

Comment:
we need to do more to determine "unhealthy", for example what if the node isn't getting unsafe blocks and not progressing?

Reply:
I agree with this but I wan't to test this on deployed testnet to get a better feel how it performs/works.  Tracked in #571